File: explorer_backend/services/clickhouse.ts
import { createClient } from "@clickhouse/client";
import { config } from "../config";

if (!config.DB_URL) throw new Error("DB_URL is not set");

export const client = createClient({
  url: config.DB_URL,
  username: config.DB_USER,
  pathname: config.DB_PATHNAME,
  password: config.DB_PASSWORD,
  database: config.DB_NAME,
});

File: nil/common/clickhouse.go
package common

import (
	"context"
	"fmt"

	"github.com/ClickHouse/clickhouse-go/v2"
)

func CreateClickHouseDbIfNotExists(ctx context.Context, dbname, user, password, endpoint string) error {
	connectionOptions := clickhouse.Options{
		Auth: clickhouse.Auth{
			Database: "system",
			Username: user,
			Password: password,
		},
		Addr: []string{endpoint},
	}
	conn, err := clickhouse.Open(&connectionOptions)
	if err != nil {
		return err
	}
	defer conn.Close()

	return conn.Exec(ctx, fmt.Sprintf("CREATE DATABASE IF NOT EXISTS %s", dbname))
}
